File: python/src/chiptime/message.py
"""Decoded message types — the lossless middle layer between wire and semantics."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from chiptime.frames import DefinitionFrame


@dataclass(frozen=True, slots=True)
class DevFieldOrigin:
    """Where a developer field came from (resolved in F6)."""

    developer_data_index: int
    field_definition_number: int
    application_id: str | None = None  # hex, from developer_data_id
    vendor: str | None = None  # manufacturer name, e.g. "stryd"
    canonical_name: str | None = None  # registry promotion for stream naming


@dataclass(frozen=True, slots=True)
class FieldValue:
    """One decoded field. `value` is scaled/normalized with sentinels → None;
    `raw` is the wire value (kept for round-trips and include_raw output)."""

    value: Any | None
    raw: Any | None = None
    units: str | None = None
    developer: DevFieldOrigin | None = None


@dataclass(frozen=True, slots=True)
class Message:
    """A decoded FIT data message, unknown-tolerant (contract #6)."""

    global_num: int
    name: str
    local_id: int
    byte_offset: int
    fields: dict[str, FieldValue] = field(default_factory=dict)
    wire: DefinitionFrame | None = None  # retained for lossless re-encoding (ADR-0006)

    def get(self, name: str) -> Any | None:
        fv = self.fields.get(name)
        return fv.value if fv is not None else None

    def get_raw(self, name: str) -> Any | None:
        fv = self.fields.get(name)
        return fv.raw if fv is not None else None

File: js/src/message.ts
/**
 * Decoded message types -- the lossless middle layer between wire and semantics.
 *
 * Twin of `python/src/chiptime/message.py`. Defined here at F33 because the frame
 * reader and the registries reference the shapes; F34's decoder is what populates
 * them.
 */

import type { DefinitionFrame } from "./frames.js";

/** Where a developer field came from (resolved in F34, mirroring Python's F6). */
export interface DevFieldOrigin {
  readonly developerDataIndex: number;
  readonly fieldDefinitionNumber: number;
  /** Hex, from `developer_data_id`. */
  readonly applicationId: string | null;
  /** Manufacturer name, e.g. "stryd". */
  readonly vendor: string | null;
  /** Registry promotion for stream naming. */
  readonly canonicalName: string | null;
}

/**
 * One decoded field.
 *
 * `value` is scaled and unit-normalized with sentinels resolved to `null`; `raw` is
 * the wire value, kept for round-trips and `includeRaw` output. A 64-bit field
 * carries `bigint` in both positions (ADR-0009 section 4).
 */
export interface FieldValue {
  readonly value: unknown;
  readonly raw: unknown;
  readonly units: string | null;
  readonly developer: DevFieldOrigin | null;
}

/** A decoded FIT data message, unknown-tolerant (contract #6). */
export interface Message {
  readonly globalNum: number;
  readonly name: string;
  readonly localId: number;
  readonly byteOffset: number;
  /**
   * Keyed by field name. A `Map` rather than an object because insertion order is
   * wire order here and the decoder walks it -- unlike the profile lookup tables,
   * where order is explicitly meaningless (ADR-0009 section 8).
   */
  readonly fields: Map<string, FieldValue>;
  /** Retained for lossless re-encoding (ADR-0006). */
  readonly wire: DefinitionFrame | null;
}

/** `msg.get(name)` in Python. Returns the scaled value, or `null` when absent. */
export function get(msg: Message, name: string): unknown {
  return msg.fields.get(name)?.value ?? null;
}

/** `msg.get_raw(name)` in Python. Returns the wire value, or `null` when absent. */
export function getRaw(msg: Message, name: string): unknown {
  return msg.fields.get(name)?.raw ?? null;
}
